gc/base/SparseVirtualMemory.cpp

```diff
@@ -0,0 +1,235 @@
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include <string.h>
+#if defined(OMR_GC_DOUBLE_MAPPING_FOR_SPARSE_HEAP_ALLOCATION)
+#include <sys/mman.h>
+#include <sys/errno.h>
+#endif /* OMR_GC_DOUBLE_MAPPING_FOR_SPARSE_HEAP_ALLOCATION */
+
+#include "omrcomp.h"
+#include "omrport.h"
+#include "omr.h"
+
+#include "EnvironmentBase.hpp"
+#include "Forge.hpp"
+#include "GCExtensionsBase.hpp"
+#include "Math.hpp"
+#include "Heap.hpp"
+#include "HeapRegionManager.hpp"
+#include "ModronAssertions.h"
+#include "SparseVirtualMemory.hpp"
+#include "SparseAddressOrderedFixedSizeDataPool.hpp"
+
+/****************************************
+ * Initialization
+ ****************************************
+ */
+
+MM_SparseVirtualMemory*
+MM_SparseVirtualMemory::newInstance(MM_EnvironmentBase* env, uint32_t memoryCategory, MM_Heap *in_heap)
+{
+	MM_SparseVirtualMemory* vmem = NULL;
+	vmem = (MM_SparseVirtualMemory*)env->getForge()->allocate(sizeof(MM_SparseVirtualMemory), OMR::GC::AllocationCategory::FIXED, OMR_GET_CALLSITE());
+
+	if (vmem) {
+		new (vmem) MM_SparseVirtualMemory(env, in_heap->getPageSize(), in_heap);
+		if (!vmem->initialize(env, memoryCategory)) {
+			vmem->kill(env);
+			vmem = NULL;
+		}
+	}
+
+	return vmem;
+}
+
+bool
+MM_SparseVirtualMemory::initialize(MM_EnvironmentBase* env, uint32_t memoryCategory)
+{
+	uintptr_t in_heap_size = (uintptr_t)_heap->getHeapTop() - (uintptr_t)_heap->getHeapBase();
+	uintptr_t maxHeapSize = _heap->getMaximumMemorySize();
+	uintptr_t regionSize = _heap->getHeapRegionManager()->getRegionSize();
+	uintptr_t regionCount = in_heap_size / regionSize;
+
+	/* Ideally this should be ceilLog2, however this is the best alternative as ceilLog2 currently does not exist */
+	uintptr_t ceilLog2 = MM_Math::floorLog2(regionCount) + 1;
+
+	uintptr_t off_heap_size = (uintptr_t)((ceilLog2 * in_heap_size) / 2.0);
+#if !defined(J9ZOS390)
+	Assert_MM_true(regionSize % _pageSize == 0);
+#endif
```

Comment:
This assertion is useless - region size and page size are defined as power of two. However there is possibility that page size is larger than region size. Can logic handle this situation properly?

Reply:
Logic can handle this for double mapping, but not currently for off-heap. I will add the same required conditions as well now for off-heap. Similar to double mapping, off-heap will now only be allowed to be enabled if the below condition is met: 

`if (!extensions->memoryManager->isLargePage(env, pagesize) || (pagesize <= extensions->getOmrVM()->_arrayletLeafSize))`